### PR TITLE
Add list-match-pattern support to match statement

### LINKED ISSUE #24291: Add list-match-pattern to match statement

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
Add list-match-pattern support to match statement with recovery logic and test cases.

Fixes #24291

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -500,7 +500,7 @@ public STNode resumeParsing(ParserRuleContext context, Object... args) {
                 return parseMatchGuard();
             case MATCH_PATTERN_START:
                 return parseMatchPattern();
-            case MATCH_PATTERN_RHS:
+            case MATCH_PATTERN_OUTER_RHS:
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:503]: Can we keep the original name?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -12492,6 +12496,100 @@ private STNode parseVarKeyword() {
...
+                break;
+            }
+            STNode matchPatternListMember = parseMatchPattern();
+            matchPatternList.add(matchPatternListMember);
+            matchPatternMemberRhs = parseListMatchPatternMemberRhs();
+
+            if (matchPatternMemberRhs != null) {
+                matchPatternList.add(matchPatternMemberRhs);
+            }
+        } while (matchPatternMemberRhs != null);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:12532]: list match pattern can also be an empty list. So a while loop would be a better option?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 3]
```diff
@@ -12492,6 +12496,100 @@ private STNode parseVarKeyword() {
...
+    private STNode parseListMatchPatternMemberRhs() {
+        return parseListMatchPatternMemberRhs(peek().kind);
+    }
+
+    private STNode parseListMatchPatternMemberRhs(SyntaxKind nextTokenKind) {
+        switch (nextTokenKind) {
+            case COMMA_TOKEN:
+                return parseComma();
+            case CLOSE_BRACKET_TOKEN:
+            case EOF_TOKEN:
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:12577]: should we add `=>` also here? What if the close bracket is missing?

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:12577]: Close bracket will recover there

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 1 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Changed

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 2 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Changed

**Pull request merged**